Sources/Rendering/LICOpenGL/LineIntegralConvolution2D/index.js

```diff
@@ -0,0 +1,563 @@
+import vtkFrameBuffer from 'vtk.js/Sources/Rendering/OpenGL/Framebuffer';
+
+import vtkLICPingPongBufferManager from 'vtk.js/Sources/Rendering/LICOpenGL/LineIntegralConvolution2D/pingpong';
+
+import { ContrastEnhanceMode } from 'vtk.js/Sources/Rendering/LICOpenGL/SurfaceLICInterface/Constants';
+
+function getVectorLookupProgram(normalize = true) {
+  // lookup the vector and normalize
+  const getNormVecSrc = `
+    vec2 getVector( vec2 vectc )\n
+      {\n
+      vec2 V = texture2D( texVectors, vectc ).xy;\n
+      // normalize if |V| not 0\n
+      float lenV = length( V );\n
+      if ( lenV > 1.0e-8 )\n
+        {\n
+        return V/lenV;\n
+        }\n
+      else\n
+        {\n
+        return vec2( 0.0, 0.0 );\n
+        }\n
+      }\n
+    `;
+
+  // lookup the vector
+  const getVecSrc = `
+    vec2 getVector( vec2 vectc )\n
+      {\n
+      return texture2D( texVectors, vectc ).xy;\n
+      }\n
+    `;
+
+  if (normalize) {
+    return getNormVecSrc;
+  }
+  return getVecSrc;
+}
+
+function vtkLineIntegralConvolution2D(publicAPI, model) {
+  model.classHierarchy.push('vtkLineIntegralConvolution2D');
+
+  publicAPI.buildAShader = (fSource) =>
+    model.openGLRenderWindow
+      .getShaderCache()
+      .readyShaderProgramArray(
+        vtkLineIntegralConvolution2D_quadVS,
+        fSource,
+        ''
+      );
+
+  publicAPI.dumpTextureValues = (
+    texture,
+    [width, height],
+    context = model.context,
+    openGLRenderWindow = model.openGLRenderWindow,
+    nComp = 4
+  ) => {
+    // To get texture values in es 2.0, we need to attach the texture to a fbo,
+    // then use glReadPixels
+    const fb = vtkFrameBuffer.newInstance();
+    const gl = context;
+    let pixels = null;
+    fb.setOpenGLRenderWindow(openGLRenderWindow);
+    fb.saveCurrentBindingsAndBuffers();
+    fb.create(width, height);
+    fb.populateFramebuffer();
+    fb.setColorBuffer(texture);
+    pixels = new Float32Array(width * height * nComp);
+    gl.readPixels(
+      0,
+      0,
+      width,
+      height,
+      nComp === 4 ? gl.RGBA : gl.RGB,
+      gl.FLOAT,
+      pixels
+    );
+    fb.restorePreviousBindingsAndBuffers();
+    return pixels;
+  };
+
+  publicAPI.getTextureMinMax = (
+    texture,
+    size,
+    context = model.context,
+    openGLRenderWindow = model.openGLRenderWindow
+  ) => {
+    const values = publicAPI.dumpTextureValues(
+      texture,
+      size,
+      context,
+      openGLRenderWindow,
+      4
+    );
+
+    let min = values[0];
+    let max = values[0];
+    for (let i = 0; i < values.length; i += 4) {
+      // Make sure the current pixel is inside the rendered geometry (g === 0)
+      if (values[i - 0 + 1] === 0.0) {
+        const val = values[i];
+        if (val < min) {
+          min = val;
+        }
+        if (val > max) {
+          max = val;
+        }
+      }
+    }
+
+    return { min, max };
+  };
+
+  publicAPI.getComponentSelectionProgram = (ids) => {
+    // swizles at 45,46
```

Comment:
what does it mean ?

Reply:
That was a debugging comment, I removed it.